WAM2layers/WAM2layers
# Add CMIP data loader

This PR took more effort than initially hoped: making all code compatible with [CF timestamps](https://github.com/Unidata/cftime) was more work than anticipated.

**Overview of changes** (in a more logical order than the commits):

- Changed input dates from `datetime.datetime` to `cftime` datetimes.
  - didn't play nice with the config's Pydantic so that code is a bit messier now :disappointed:  @Peter9192 
- All timestamps internally are now cftime based, and date ranges are `xarray.CFTimeIndex` objects
- Instead of a simple loop and [checking the hour to see if the data covers a new day](https://github.com/WAM2layers/WAM2layers/blob/ffee8b844af92a991a9cbe7d5f70a408f25498e6/src/wam2layers/preprocessing/shared.py#L94), the preprocessing dates are [split using a groupby operation](https://github.com/WAM2layers/WAM2layers/blob/ddd457dbce507ab642bf81ce707f0e63cd240296/src/wam2layers/preprocessing/shared.py#L113-L114).
  - This allowed adding an experimental parallel preprocessor (speeds up preprocessing on my machine 4-8x, so it made for much faster testing)
  - The parallel preprocessing currently only runs if explicitly enabled in the config file
- Added caching to the opening of netCDF files for a minor speed boost (using `@cache` and `@lru_cache`)
- Added CMIP preprocessor (that works with the input data listed below), you now can do `wam2layers preprocess cmip`

<hr>

**Current input data requirements**

Note that this data loader is made for the specific files from MPI which are on a 3hr and 6hr frequency: this has not been generalized yet.

If you use variables from the following tables, the preprocessor should work for other climate models as well:
- mip table "3hr" vars: `hfls`, `pr`
- mip table "6hrPlevPt" vars: `hus`, `huss`, `ua`, `uas`, `va`, `vas`
- mip table "6hrLev" var: `ps`

<details>
<summary>example config</summary>

```yaml
# General
preprocessed_data_folder: ./preprocessed_data
output_folder: ./output_data_fw
tracking_direction: forward
calendar: standard

# Preprocessing
filename_template: ./input_data/{variable}_*.nc
preprocess_start_date: "2030-03-01T00:00"
preprocess_end_date: "2030-08-01T00:00"
parallel_preprocess: True
level_type: pressure_levels
levels: [5000, 25000, 50000, 60000, 70000, 85000, 92500]

# Tracking
tracking_domain: null  # use full domain of preprocessed data
# tracking_domain: [0, 50, 10, 55]  # use subdomain of preprocessed data; west south east north
tracking_start_date: "2030-03-14T00:00"
tracking_end_date: "2030-08-01T00:00"

# Tagging
tagging_region: [-10, 35, 40, 60]  # alternative to mask file; west south east north
tagging_start_date: "2030-03-14T00:00"
tagging_end_date: "2030-05-14T00:00"

input_frequency: "6h"
timestep: 600  # timestep in seconds
output_frequency: "1D"
restart: False
periodic_boundary: true
kvf: 3
```

</details>

<hr>

**How to test/review:**

- Run download script (`scripts/download_MPI_cmip_data.sh`)
- Setup config file
- do `wam2layers preprocess cmip config_file.yaml`
  - note that it'll use the parallel preprocessor so it'll take more RAM/cpu
- do `wam2layers track config_file.yaml`

<hr>

**Todo:**

- [x] Check if results make sense (@Imme1992 ?)
- [x] Make sure output files have right calendar system
- [x] Open issues for points below;

Todo later:
- [ ] Make code more robust (fewer assumptions, i.e. input data freq check / tie vars to mip tables)
- [ ] Add regression tests
  - [ ] Example CMIP config
  - [ ] CMIP example dataset
  - [ ] Which variables from which mip tables are required
